From 6a2976d58e99da5e3a0d9694d0acd3d562d4fa72 Mon Sep 17 00:00:00 2001
From: hariprasad <hkelam@marvell.com>
Date: Thu, 11 Jul 2019 15:51:41 +0530
Subject: [PATCH 338/386] octeontx2-af: Delete range of mcam entries

Extend npc_delete_flow functionality to delete range of mcam
entries.This feature will be used to delete rules associated with
specific filters (ntuple,unicast etc).

Change-Id: Ia8e8713c5ccc77f5abe654a1c7c19eb342affe65
Signed-off-by: hariprasad <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/12420
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/mbox.h   |  2 ++
 .../net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c | 14 ++++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/mbox.h b/drivers/net/ethernet/marvell/octeontx2/af/mbox.h
index 83d3838be83e..0d4383d2e50b 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/mbox.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/mbox.h
@@ -1286,6 +1286,8 @@ struct npc_install_flow_rsp {
 struct npc_delete_flow_req {
 	struct mbox_msghdr hdr;
 	u16 entry;
+	u16 start;/*Disable range of entries */
+	u16 end;
 	u8 all; /* PF + VFs */
 };
 
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
index c331aad26e7f..eaf39b4da14a 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
@@ -961,6 +961,20 @@ int rvu_mbox_handler_npc_delete_flow(struct rvu *rvu,
 	u16 pcifunc = req->hdr.pcifunc;
 	int err;
 
+	if (req->end) {
+		list_for_each_entry_safe(iter, tmp, &mcam->mcam_rules, list) {
+			if (iter->owner == pcifunc &&
+			    iter->entry >= req->start &&
+			    iter->entry <= req->end) {
+				err = npc_delete_flow(rvu, iter->entry,
+						      pcifunc);
+				if (err)
+					return err;
+			}
+		}
+		return 0;
+	}
+
 	if (!req->all)
 		return npc_delete_flow(rvu, req->entry, pcifunc);
 
-- 
2.17.1

